src/DurableTask.Core/TaskOrchestrationContext.cs

```
@@ -249,7 +249,7 @@ public override async Task<T> CreateTimer<T>(DateTime fireAt, T state, Cancellat
             var createTimerOrchestratorAction = new CreateTimerOrchestratorAction
             {
                 Id = id,
-                FireAt = fireAt,
+                FireAt = fireAt.ToUniversalTime(),
```

We have to be a bit careful here since we're changing behavior in DTFx.Core, which means it affects all backends, not just Azure Storage. For example - I'm not sure if Netherite performs an implicit conversion to universal time, or if it assumes all time is UTC.

This makes me think it may be safer to throw if the time is not in UTC. That way, it'll be simpler to predict the behavior. Not sure, would appreciate more thoughts

---

Yes, I thought about it, and I'm still on the fence. I realize it will impact all backends, whether we throw or silently convert, and this is why it is intended to go to the next major version of DT.Core. I agree with the benefits of throwing an exception (after all, this is what the Azure SDK owners chose to do), but adhering to the old behavior would make the 
 migration from the previous version easier (for example, the original issue would not even be raised).

---

Another option would be not to do the conversion here (DTFx.Core), but to do it in DTFx.AzureStorage, at some point before we save state in the history/queue. By looking at the original trace, you may be able to spot where that conversion should be.

---

> Another option would be not to do the conversion here (DTFx.Core), but to do it in DTFx.AzureStorage, at some point before we save state in the history/queue. By looking at the original trace, you may be able to spot where that conversion should be.

Yes, this is an option. What I didn't like about it is that it will require checking every event type and explicit type casting, probably around here: https://github.com/Azure/durabletask/blob/d4b98d46158e470c6b769601cd76a86d9bb6d096/src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs#L1020
However, we already do type-specific processing there, so this patch will not make it significantly worse.

Ok, this will limit the impact to the AzureStorage backend. Now, the question is: do we _want_ to limit this to the Azure Storage backend? Regardless of the backend and the current behavior, if the customer is passing non-UTC time to `CreateTimer`, chances are they may be confused. Given that UTC is expected in many other parts, including DT.Core (such as context.CurrentUtcDateTime), do we want them to continue mixing-and-matching? Accidentally, we are bumping the DT.Core version now, so this is a good opportunity to make an intentional breaking change.

---

Unless we have a concrete (non-theoretical) reason to change DurableTask.Core, I suggest we avoid it and only change DurableTask.AzureStorage. I personally think it's fine if users want to pass local times to our APIs. I consider the fact that we internally save those timestamps as UTC as an internal implementation detail. Since APIs like `context.CurrentUtcDateTime` already make it clear that the timestamp returned is in UTC, I don't think we need to be concerned about confusing them.

---

Reverted DurableTask.Core changes, so this fix is limited to DurableTask.AzureStorage now.